File: lib/services/authentication_service.js
const Services = require('./service');
const { Login } = require('./../models/login');
const { Logout } = require('./../models/logout');

/**
 * Handles login and logout of a user, allowing us to get an authTokenKey to 
 * be used to call other services.
 */
class AuthenticationService extends Services {
  constructor(endPoint) {    
    super(endPoint);
  }

  /**
   * Log the user in
   * @param tenant The user's tenant (organization name)
   * @param username The user's username
   * @param password The user's password
   * @returns Login
   */
  async login(tenant, username, password) {
    let result = await this.executeOperation("/user/login", {
      username: username,
      password: password,
      authTokenKey: "RANDOMCODE",
      tenant: tenant,
      clientInfo: "mobile"
    });

    if (result) {
      return new Login(result.content);
    } else {
      throw Error(`failed to login in with user ${username}`);
    }
  }

  /**
   * Log the user out
   * @param userId The user's userId
   * @returns Logout
   */
  async logout(authTokenKey, tenant, userId) {
    let result = await this.executeOperation("/user/logout", {
      userId: userId,
      authTokenKey: authTokenKey,
      tenant: tenant,
      clientInfo: "mobile"
    });

    if (result) {
      return new Logout(result.content);
    } else {
      throw Error(`failed to logout for userId ${userId}`);
    }
  }
}

module.exports = AuthenticationService;

File: libraries/java/src/main/java/com/qmplus/v3/api/services/AuthenticationService.java
package com.qmplus.v3.api.services;

import com.fasterxml.jackson.core.type.TypeReference;
import com.qmplus.v3.api.models.request.LoginRequest;
import com.qmplus.v3.api.models.request.LogoutRequest;
import com.qmplus.v3.api.models.response.LoginResponse;
import com.qmplus.v3.api.models.response.LogoutResponse;
import com.qmplus.v3.api.models.response.ResponseWrapper;
import com.qmplus.v3.api.models.util.RestConstants;

import java.io.IOException;

public class AuthenticationService extends BaseService {
  AuthenticationService() {
    super();
  }

  AuthenticationService(String endPoint) {
    super(endPoint);
  }

  LoginResponse login(
      String tenant, String username, String password, String clientInfo
  ) throws IOException {
    LoginRequest request = new LoginRequest();

    // Trigger the default random key
    request.setAuthTokenKey(RestConstants.DEFAULT_AUTH_TOKEN_KEY);
    request.setTenant(tenant);
    request.setUsername(username);
    request.setPassword(password);
    request.setClientInfo(clientInfo != null ? clientInfo : "");

    // Execute the operation
    ResponseWrapper<LoginResponse> response
        = executeOperation(new TypeReference<ResponseWrapper<LoginResponse>>() {},"user/login", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }

  LogoutResponse logout(
      String authTokenKey, String tenant, Integer userId, String clientInfo
  ) throws IOException {
    LogoutRequest request = new LogoutRequest();

    // Trigger the default random key
    request.setAuthTokenKey(authTokenKey);
    request.setTenant(tenant);
    request.setUserId(userId);
    request.setClientInfo(clientInfo != null ? clientInfo : "");

    // Execute the operation
    ResponseWrapper<LogoutResponse> response
        = executeOperation(new TypeReference<ResponseWrapper<LogoutResponse>>() {},"user/logout", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }
}
